token/js/src/actions/amountToUiAmount.ts

```diff
@@ -28,3 +31,189 @@ export async function amountToUiAmount(
+    amount: string,
+    decimals: number,
+    currentTimestamp: number, // in seconds
+    lastUpdateTimestamp: number,
+    initializationTimestamp: number,
+    preUpdateAverageRate: number,
+    currentRate: number,
+): string {
+    Decimal.set({ toExpPos: 24, toExpNeg: -24 })
+
+    // Calculate pre-update exponent
+    const preUpdateTimespan = new Decimal(lastUpdateTimestamp).minus(initializationTimestamp);
+    const preUpdateNumerator = new Decimal(preUpdateAverageRate).times(preUpdateTimespan);
+    const preUpdateExponent = preUpdateNumerator.div(new Decimal(SECONDS_PER_YEAR).times(ONE_IN_BASIS_POINTS));
+    const preUpdateExp = preUpdateExponent.exp();
+
+    // Calculate post-update exponent
+    const postUpdateTimespan = new Decimal(currentTimestamp).minus(lastUpdateTimestamp);
+    const postUpdateNumerator = new Decimal(currentRate).times(postUpdateTimespan);
+    const postUpdateExponent = postUpdateNumerator.div(new Decimal(SECONDS_PER_YEAR).times(ONE_IN_BASIS_POINTS));
+    const postUpdateExp = postUpdateExponent.exp();
+
+    // Calculate total scale
+    const totalScale = preUpdateExp.times(postUpdateExp);
+
+    // Calculate scaled amount with interest rounded down to the nearest unit
+    const decimalsFactor = new Decimal(10).pow(decimals);
+    const scaledAmountWithInterest = new Decimal(amount).times(totalScale).div(decimalsFactor).toDecimalPlaces(decimals, Decimal.ROUND_DOWN);
+
+    return scaledAmountWithInterest.toString();
+}
+
+/**
+ * Convert amount to UiAmount for a mint with interest bearing extension without simulating a transaction
+ * This implements the same logic as the CPI instruction available in /token/program-2022/src/extension/interest_bearing_mint/mod.rs
+ *
+ * @param connection     Connection to use
+ * @param mint           Mint for the account
+ * @param amount         Amount of tokens to be converted to Ui Amount
+ * @param programId      SPL Token program account (default: TOKEN_PROGRAM_ID)
+ *
+ * @return Ui Amount generated
+ */
+export async function amountToUiAmountForMintWithoutSimulation(
+    connection: Connection,
+    mint: PublicKey,
+    amount: string,
+    programId = TOKEN_PROGRAM_ID,
+): Promise<string> {
+    Decimal.set({ toExpPos: 24, toExpNeg: -24 })
+    const mintInfo = await getMint(connection, mint, 'confirmed', programId);
+    const amountDecimal = new Decimal(amount.toString());
+    const decimalsFactor = new Decimal(10).pow(mintInfo.decimals);
+
+    if (programId.equals(TOKEN_PROGRAM_ID)) {
+        console.log('amountDecimal', amountDecimal.toString(), 'mintInfo', mintInfo);
+        return amountDecimal.div(decimalsFactor).toString();
+    }
+
+    const interestBearingMintConfigState = getInterestBearingMintConfigState(mintInfo);
+    if (!interestBearingMintConfigState) {
+        return amountDecimal.div(decimalsFactor).toString();
+    }
+
+    const currentTime = Math.floor(Date.now() / 1000); // Convert to seconds
```

Comment:
Up to you, but it would be more correct to use the unix timestamp reported by the sysvar clock: https://docs.anza.xyz/runtime/sysvars/#clock, but that would also require fetching another account

Reply:
sure. it makes mocks a bit more complicated but let's give it a try